aten/src/ATen/core/SingletonSymNodeImpl.h

```diff
@@ -34,12 +35,40 @@ namespace c10 {
 // During tracing the strides of the outputs need to be a function of the size
 // and strides of the inputs so it is important that SingletonSymNode itself is
 // able to express this.
+//
+// NOTE [ SingletonVariant ]
+//
+// Currently, if SingletonSymNodeType::CPP is passed, that means that the
+// singleton is only meant to be used to ferry nested_tensor_size metadata
+// from forward to use in backward. In this case we set `val`, `coeff` etc
+// to bogus values and make sure to error if they are accessed.
+enum class SingletonVariant { PYTHON, CPP };
+
+constexpr c10::DispatchKeySet py_singleton_ks({c10::DispatchKey::Python, c10::DispatchKey::PythonTLSSnapshot});
+constexpr c10::DispatchKeySet cpp_singleton_ks({c10::DispatchKey::NestedTensor});
+
 class TORCH_API SingletonSymNodeImpl : public SymNodeImpl {
  public:
   // CAUTION: you should probably not be constructing these directly; please
-  // the higher-level API in python instead (TODO: actually introduce that).
-  explicit SingletonSymNodeImpl(int64_t val, int64_t coeff)
-      : val_(val), coeff_(coeff) {}
+  // the higher-level API in python instead.
+  explicit SingletonSymNodeImpl(
+      int64_t val,
+      int64_t coeff,
+      at::Tensor vec,
+      int64_t sum_vec,
+      SingletonVariant type)
+      : val_(val), coeff_(coeff), vec_(std::move(vec)), sum_vec_(sum_vec), type_(type) {
+    // See NOTE [ SingletonVariant ]
+    if (type == SingletonVariant::PYTHON) {
+      key_set_ = py_singleton_ks;
```

Comment:
Since singleton has gone full nested tensor I would really like to see the rename sooner rather than later. It gets harder to rename the more and more references you add to singleton.